### PR TITLE
Typescript support

### PR DESCRIPTION
**Pull request by imeredith**
This adds typescript support to blueocean. Its still somewhat a WIP, but enough to be merged I believe.

### CODE REVIEW

File: blueocean-dashboard/package.json [hunk 1]
```diff
@@ -89,5 +96,34 @@
       "react-router@any",
       "redux-thunk@any"
     ]
+  },
+  "jest": {
```
Comment (kzantow) [inline blueocean-dashboard/package.json]: might be good to have this config in the external jest config file `jest.config.js`

File: blueocean-dashboard/src/main/js/Dashboard.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+interface DashBoardContext {
+    pipelineService?: object,
+    activityService?: object,
+    params?: object,
+    location?: object
+}
+class Dashboard extends React.Component<DashboardProps, any> {
+    static childContextTypes = {
```
Comment (kzantow) [inline blueocean-dashboard/src/main/js/Dashboard.tsx]: erhmergerhd... i have a better solution for this, but not part of this pr

File: blueocean-dashboard/package.json (hunk 1 above)
Comment (imeredith) [reply blueocean-dashboard/package.json]: That is true, i would have to update the js-builder-jest plugin though. Which i might do in another pr

File: bin/rebuildShrinkwrap.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * We will prune and install BEFORE we install the requested version to make sure that
+ * shrinkwrap will update correctly everytime. We further do an optional `mvn install`
+ * to publish the new hpi to the local .m2 repository
+ *
+ * Usage: bin/cleanInstall.js [[@scope/]package@version] [dev] [mvn]
+ *
+ * dev will install a devDependency
+ * mvn will run the optional `mvn install` afterwards
+ */
```
Comment (sophistifunk) [inline bin/rebuildShrinkwrap.js]: Is this meant to replace `bin/cleanInstall.js` ?

File: blueocean-core-js/src/js/fetch.ts [hunk 4]
```diff
@@ -59,26 +76,24 @@ export const FetchFunctions = {
...
      * This should only be used if not using fetch or fetchJson.
      */
-    checkStatus(response) {
+    checkStatus(response: Response): Response {
         if (response.status >= 300 || response.status < 200) {
             const message = `fetch failed: ${response.status} for ${response.url}`;
-            const error = new Error(message);
-            error.response = response;
-            throw error;
+            throw {message , response};
```
Comment (sophistifunk) [inline blueocean-core-js/src/js/fetch.ts]: We should be throwing `Error` instances, not objects. The stack trace comes from the object creation, not the throw statement.

File: blueocean-core-js/src/js/fetch.ts [hunk 5]
```diff
@@ -101,7 +116,7 @@ export const FetchFunctions = {
      * Usage:
      * FetchUtils.fetch(..).then(FetchUtils.parseJSON)
      */
-    parseJSON(response) {
+    parseJSON<T>(response: Response): Promise<T> {
```
Comment (sophistifunk) [inline blueocean-core-js/src/js/fetch.ts]: Delicious typed promises 😍

File: blueocean-core-js/test/js/fetch-spec.js [hunk 6]
```diff
@@ -67,6 +67,7 @@ describe('Fetch', () => {
                             assert.fail(null, null, 'should not call success handler');
                         },
                         error => {
+                            console.log('XXXXX:' ,error);
```
Comment (sophistifunk) [inline blueocean-core-js/test/js/fetch-spec.js]: Leftover debugging, or is this meant to be here?

File: blueocean-dashboard/src/main/js/Dashboard.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import { pipelineService, activityService } from '@jenkins-cd/blueocean-core-js';
+import navState from './DashboardNavState';
+
+interface Props {
+    params: object
+    children?: React.ReactNode
+    location: object
+}
+
+interface DashBoardContext {
```
Comment (sophistifunk) [inline blueocean-dashboard/src/main/js/Dashboard.tsx]: Yay!

But that capital B should probably be a normal b :)

File: blueocean-pipeline-editor/tsconfig.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (sophistifunk) [inline blueocean-pipeline-editor/tsconfig.json]: Do we have anything in place yet that will help us keep these files the same?

**Comment by sophistifunk** [comment]
Looks great, can't wait to leverage this when I get back to pipeline results!

File: bin/rebuildShrinkwrap.js (hunk 3 above)
Comment (imeredith) [reply bin/rebuildShrinkwrap.js]: No, i need to fix up that file

File: blueocean-core-js/test/js/fetch-spec.js (hunk 6 above)
Comment (imeredith) [reply blueocean-core-js/test/js/fetch-spec.js]: No cheers

File: blueocean-pipeline-editor/tsconfig.json (hunk 8 above)
Comment (imeredith) [reply blueocean-pipeline-editor/tsconfig.json]: No.  I tried to symlink them but tooling seems to not like that.
